pkg/operator/sync.go

```diff
@@ -78,7 +79,21 @@ func (c *csiDriverOperator) syncDeployment(instance *v1alpha1.Driver) (*appsv1.D
 		forceRollout = true
 	}
 
-	deploy, _, err := resourceapply.ApplyDeployment(
+	existingDeployment, err := c.deploymentInformer.Lister().Deployments(deploy.Namespace).Get(deploy.Name)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		// Fallthrough on IsNotFound - ApplyDeployment will get the Deployment from API
+		// server and create it if it does not exists. forceRollout = true would redeploy
+		// when our informer is stale.
+	} else {
+		if c.containerImagesNeedsUpdate(&existingDeployment.Spec.Template.Spec, &deploy.Spec.Template.Spec) {
+			forceRollout = true
+		}
+	}
```

Comment:
Have you considered doing this in library-go's `ApplyDeployment()`? I think it'd be cleaner, as `ApplyDeployment` needs to `Get` the object anyway.

Updating the object when there's a new container image seems to be desired, so it shouldn't be a problem for other users of `ApplyDeployment()`?

Reply:
To clarify, perhaps even better would be to compare the `Spec`, something like:
```diff
diff --git a/pkg/operator/resource/resourceapply/apps.go b/pkg/operator/resource/resourceapply/apps.go
index 253b17ef..1a8ecb0a 100644
--- a/pkg/operator/resource/resourceapply/apps.go
+++ b/pkg/operator/resource/resourceapply/apps.go
 	if !*modified && existingCopy.ObjectMeta.Generation == expectedGeneration && !forceRollout {
-		return existingCopy, false, nil
+		if equality.Semantic.DeepEqual(required.Spec, existingCopy.Spec) {
+			return existingCopy, false, nil
+		}
 	}
```

Reply:
This has bothered me for awhile but none of the `Apply*` functions in library-go check for spec. Are you proposing that we patch it in library-go?

Reply:
IMO you can't check Spec, some fields gets defaulted by API server. That's IMO the reason why Generation was introduced.

And each team will have different opinion what should be checked - now we want conainer images, but someone will want also cmdline arguments (that's where the log level is) and whatnot.

It would be better to change resourceapply functions to accept existing object, if it is known.

Reply:
@deads2k, are we missing anything? This PR tries to re-deploy operand (DaemonSet + Deployment) when the operator is restarted with env. variables pointing to new images. Someone must have it solved already. How other operators detect this?

Reply:
library-go's `ApplyDeployment` uses generation, not spec to determine whether it needs to be applied.  See the check here: https://github.com/openshift/library-go/blob/master/pkg/operator/resource/resourceapply/apps.go#L39-L42

The recommended .status for operators keep track of the last applied generation.  You can see usage here: https://github.com/openshift/cluster-openshift-apiserver-operator/blob/1bc63aca70a3258876befa7db6aeb8e6e05cee87/pkg/operator/workload/workload_openshiftapiserver_v311_00_sync.go#L396

Reply:
@deads2k Is there a point to tracking the generation if `ApplyDeployment` sets the deployment spec as an annotation?